### PR TITLE
Reland "Move tensor grouping to ATen"

### PR DESCRIPTION
**Pull request by malfet**
This is a reland of https://github.com/pytorch/pytorch/pull/100007 with a build fix for Windows debug builds.
`at::native::ParamsHash` only works on structs with standard layout, but `std::string` isn't one in Visual C++ debug builds, which one can easily verified by running something like:
```cpp
#define _DEBUG
#include <type_traits>
#include <string>
static_assert(std::is_standard_layout_v<std::string>, "Oh noes");
```
If above conditon is not met, instead of printing a static_assert output, VC++ raises a very cryptic compilation errors,  see https://github.com/pytorch/pytorch/pull/100007#discussion_r1227116292 for more detail.

Also, using `std::hash` for string should result in a faster hash function.

(cherry picked from commit 74b7a6c75e698378882d30958908073407f97fb3)

### <samp>🤖 Generated by Copilot at 5914771</samp>

This pull request introduces a new function `_group_tensors_by_device_and_dtype` that can group tensors by their device and dtype, and updates the `foreach` utilities and several optimizers to use this function. The goal is to improve the performance, readability, and compatibility of the code that handles tensors with different properties. The pull request also adds a test case and type annotations for the new function, and some error checks for the `fused` argument in Adam and AdamW.

### CODE REVIEW

File: torch/csrc/Module.cpp [hunk 1]
```diff
@@ -1723,6 +1724,52 @@ Call this whenever a new thread is created in order to propagate values from
...
+  // standard layout, but std::string isn't one in Visual C++ debug builds,
+  // which one can easily verify by running something like
+  //   #define _DEBUG
+  //   #include <type_traits>
+  //   #include <string>
+  //   static_assert(std::is_standard_layout_v<std::string>, "Oh noes");
+  // But if this conditon is not met, VC++ raises a very cryptic compilation
+  // errors See
+  // https://github.com/pytorch/pytorch/pull/100007#discussion_r1227116292 for
+  // more detail
```
Comment (janeyx99) [inline torch/csrc/Module.cpp:1745]:
```suggestion
  // But if this condition is not met, VC++ raises a very cryptic compilation
  // error. See
  // https://github.com/pytorch/pytorch/pull/100007#discussion_r1227116292 for
  // more detail
```
nit

File: torch/csrc/Module.cpp [hunk 2]
```diff
@@ -1723,6 +1724,52 @@ Call this whenever a new thread is created in order to propagate values from
...
+  struct _DeviceDtypeHasher {
+    std::size_t operator()(const _DeviceDtypeKey& k) const noexcept {
+      static at::native::ParamsHash<at::Device> device_hasher;
+      static std::hash<std::string> string_hasher;
+      return device_hasher(k.first) ^ string_hasher(k.second);
```
Comment (janeyx99) [inline torch/csrc/Module.cpp:1750]: Just curious, why are we xoring/does it matter?

**Review by janeyx99 — APPROVED** [review]
A big big thank you to @malfet for digging so deep into this and coming to a resolution.

File: torch/csrc/Module.cpp (hunk 2 above)
Comment (malfet) [reply torch/csrc/Module.cpp:1750]: Great question!
- `xor` is fast, balanced and entropy preserving binary op, so it's very useful for joining hashes.
- But does it matter? Not really, no, as those maps are presumably quite small.

File: torch/optim/adam.py [hunk 3]
```diff
@@ -49,6 +49,15 @@ def __init__(self, params, lr=1e-3, betas=(0.9, 0.999), eps=1e-8,
                                    f"supported devices: {fused_supported_devices}.")
             if foreach:
                 raise RuntimeError("`fused` and `foreach` cannot be `True` together.")
+            # TODO(crcrpar): [low prec params & their higher prec copy]
+            # Suppor AMP with FP16/BF16 model params which would need
```
Comment (kit1980) [inline torch/optim/adam.py:53]:
```suggestion
            # Support AMP with FP16/BF16 model params which would need
```

File: torch/optim/adamw.py [hunk 4]
```diff
@@ -67,6 +67,15 @@ def __init__(
                                    f"supported devices: {fused_supported_devices}.")
             if foreach:
                 raise RuntimeError("`fused` and `foreach` cannot be `True` together.")
+            # TODO(crcrpar): [low prec params & their higher prec copy]
+            # Suppor AMP with FP16/BF16 model params which would need
```
Comment (kit1980) [inline torch/optim/adamw.py:71]:
```suggestion
            # Support AMP with FP16/BF16 model params which would need
```

File: torch/csrc/Module.cpp [hunk 5]
```diff
@@ -1723,6 +1724,52 @@ Call this whenever a new thread is created in order to propagate values from
...
+  using _DeviceDtypeKey = std::pair<at::Device, std::string>;
+  // Custom hasher is necessary to make unordered_map compilable for Windows
+  // debug targets As `at::native::ParamsHash` only works on structs with
+  // standard layout, but std::string isn't one in Visual C++ debug builds,
+  // which one can easily verify by running something like
+  //   #define _DEBUG
+  //   #include <type_traits>
+  //   #include <string>
+  //   static_assert(std::is_standard_layout_v<std::string>, "Oh noes");
+  // But if this condition is not met, VC++ raises a very cryptic compilation
```
Comment (malfet) [inline torch/csrc/Module.cpp:1742]:
```suggestion
  // If above condition is not met, VC++ raises a very cryptic compilation
```

File: torch/csrc/Module.cpp [hunk 6]
```diff
@@ -1723,6 +1724,52 @@ Call this whenever a new thread is created in order to propagate values from
...
+  // Currently I see the second item of the key is displayed as
+  // e.g. `torch._C._te.ScalarType at 0x7fcf318adab0`
+  // I thought adding an appropriate type_caster of `at::ScalarType` to
+  // torch/csrc/pybind.h` would solve this but it caused segmentation fault in
+  // my environment.
+  using _DeviceDtypeKey = std::pair<at::Device, std::string>;
+  // Custom hasher is necessary to make unordered_map compilable for Windows
+  // debug targets As `at::native::ParamsHash` only works on structs with
+  // standard layout, but std::string isn't one in Visual C++ debug builds,
+  // which one can easily verify by running something like
```
Comment (malfet) [inline torch/csrc/Module.cpp:1737]:
```suggestion
  // which one can easily verify by running something like:
```

File: torch/csrc/Module.cpp [hunk 7]
```diff
@@ -1723,6 +1724,52 @@ Call this whenever a new thread is created in order to propagate values from
...
 
+  // FIXME(crcrpar): Better to have `at::ScalarType` get mapped to `torch.dtype`
+  // Currently I see the second item of the key is displayed as
+  // e.g. `torch._C._te.ScalarType at 0x7fcf318adab0`
+  // I thought adding an appropriate type_caster of `at::ScalarType` to
+  // torch/csrc/pybind.h` would solve this but it caused segmentation fault in
+  // my environment.
+  using _DeviceDtypeKey = std::pair<at::Device, std::string>;
+  // Custom hasher is necessary to make unordered_map compilable for Windows
+  // debug targets As `at::native::ParamsHash` only works on structs with
```
Comment (malfet) [inline torch/csrc/Module.cpp:1735]:
```suggestion
  // debug targets. As `at::native::ParamsHash` only works on structs with
```

**Comment by crcrpar** [comment]
thank you very much @malfet

**Pull request closed without merging**